### ISSUE TITLE
Cyclic dependencies between components

### ISSUE THREAD

**Issue by pyricau** [google/dagger#311]
Let's say I have an isolated library that has inputs and outputs (it can build some things, it needs other things to do so). Components & components dependencies are great to express those:

```java
public interface UserLibDependencies {
  HttpClient httpClient();
}

@Component(dependencies = UserLibDependencies.class, modules = UserModule.class)
public interface UserLibComponent {
  User user();
}
```

Now, let's say I want to use this library in my app top level component, which is able to build an `HttpClient`, and will need a `User`.

```java
@Component(dependencies = UserLibComponent.class, modules = AppModule.class)
public interface AppComponent implements UserLibDependencies {
}
```

Obviously, you see the problem: this can't work because both components need each other to be built, the Dagger2 generated builders won't let us create partially built components.

I wonder if Dagger should have an option to build a component with a missing dependency that can be set after the component has been created.

In the meantime, this problem can be solved by building a proxy that acts as a partially built component:

```java
/**
 * Holds a lazy proxy that will delegate all method calls to the real implementation of the
 * interface this is parameterized with. Useful to solve cyclic dependencies between component
 * dependencies (for which we cannot use providers).
 */
public interface DelegateProxy<T> {

  void setDelegate(T delegate);

  T proxy();

  class Helper {

    public static <T> DelegateProxy<T> proxyDelegate(Class<T> interfaceClass) {
      return new DelegateInvocationHandler<>(interfaceClass);
    }

    private static class DelegateInvocationHandler<T>
        implements InvocationHandler, DelegateProxy<T> {

      private final Class<T> delegateInterface;
      private final T proxy;

      private T delegate;

      DelegateInvocationHandler(Class<T> delegateInterface) {
        checkState(delegateInterface.isInterface());
        this.delegateInterface = nonNull(delegateInterface, "delegateInterface");
        //noinspection unchecked
        proxy = (T) Proxy.newProxyInstance(delegateInterface.getClassLoader(),
            new Class<?>[] { delegateInterface }, this);
      }

      @Override public void setDelegate(T delegate) {
        if (this.delegate != null) {
          throw new UnsupportedOperationException("Can't set delegate twice");
        }
        if (!delegateInterface.isInstance(delegate)) {
          throw new IllegalArgumentException(
              delegate + " is not an instance of " + delegateInterface.getName());
        }
        this.delegate = delegate;
      }

      @Override public T proxy() {
        return proxy;
      }

      @Override public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
        return method.invoke(delegate, args);
      }
    }
  }
}
```

Usage:

```java
DelegateProxy<UserLibDependencies> delegate = DelegateProxy.Helper.proxyDelegate(UserLibDependencies.class);
UserLibComponent userComponent = DaggerUserLibComponent.builder().userLibDependencies(delegate.proxy).build();

AppComponent appComponent = DaggerAppComponent.builder().userLibComponent(userComponent).build();

delegate.setDelegate(appComponent);
```

**Comment by gk5885** [comment]
I think that I disagree with the premise of the feature request.  While you _can_ use components to build libraries, in this case, I think that modules are the better way to express, well, modularity.  It seems that from your example, you need an `HttpClient` to fetch a `User`, which you could express by composing the `UserModule` and whichever module binds the `HttpClient` rather than using an interface and component deps.  If you don't have the need to compose them separately, you can use `includes` to compose them, but I'm guessing from the fact that you have introduced the interface that you want to be able to swap out which `HttpClient` is bound, so you can just add both modules directly to the component and rely on the compile-time validation to ensure that you have all of your deps satisfied.

**Comment by pyricau** [comment]
The main reason is that because I don't think libraries should leak implementation details such has "how to do dependency injection". I don't think it's ok for a library to provide a module to be included. It also means that your library gets no graph validation, and no clear expression of its input / outputs.

Arguably, a library could have the components I declared here just for graph validation purposes and clearly identifying bounds, and then expose the module to include directly. I'm still not a fan (because, again, this requires that you use the same DI lib / version etc on both sides).

I'm going to close this issue, I'm ok with disagreeing, since I have a solution anyway.

**Comment by gk5885** [comment]
> I don't think libraries should leak implementation details such has "how to do dependency injection"

I completely agree with the sentiment, but anything having to do with scope completely breaks that goal.  `@Scope` is defined as a signal to the injector and the injector is quite specifically how you do DI.  So, if you use Dagger as purely an implementation detail in a library, you would either:

*   Define a scope that somehow models a lifecycle that is controlled and used exclusively by your library
*   Strictly use unscoped bindings, but if you absolutely have to, manage instances with broader instance management tools like enums, `ThreadLocal`, etc.

Otherwise you have to do a fair bit of work to make sure that two separately interact in a way that the separate injectors are still honoring the spirit of the scopes that you've used.  E.g., make sure that two libraries don't each bind their own `@Singleton` `HttpClient`.

As far as the library-level validation goes, there isn't any good out-of-the-box way to declare a "module interface".  It has been discussed, but always feels a bit awkward for all the same reasons that Dagger 1 module-level validation did -- it takes a lot of information to express your intent and everybody seems to just say screw it and turn it off anyway.  That said, if it really is something you want to enforce, you can make an interface for your inputs that you use as a component dep and an interface for your outputs that you declare as a component that installs your module.  Now you have a compile-time test that validates that relationship even if it's never actually a component that you use.

So, to recap, the primary issue with components in libraries and component deps in general is that it's too easy to violate the logical constraints of a scope.  We're not even guaranteed that a component dep is implemented by Dagger so it's certainly hard to enforce that they're being used in a consistent way.  If we had a good solution to that problem, component deps might be a more universal solution, but until then library modules seem to be the more robust solution, IMO.

**Comment by pyricau** [comment]
thx for the details.
